plasmapy/formulary/parameters.py

```diff
@@ -96,64 +97,101 @@ def _grab_charge(ion: Particle, z_mean=None):
+        ``'D+'`` for deuterium, or ``'He-4 +1'`` for singly ionized helium-4).
+
+    z_ratio : `int`, `float`, optional
+        The ratio of the integer charges coresponding to the plasma species
+        represted by ``density`` and the ``particle``.  For example, if the
+        given ``density`` is and electron density and ``particle`` is doubly
+        ionized ``He``, then ``z_ratio = -0.5``.  Defalut is ``1``.
 
     Raises
     ------
-    ValueError
-        If the `density` has units inconvertible to either a particle density
-        or a mass density, or if you pass in a number density without a particle.
+    `~astropy.units.UnitTypeError`
+        If the ``density`` does not have units equivalent to a number denisty
+        or mass density.
+
+    `TypeError`
+        If ``density`` is not of type `~astropy.units.Quantity`, or convertible.
+
+    `TypeError`
+        If ``particle`` is not of type or convertible to
+        `~plamsapy.particles.Particle`.
+
+    `TypeError`
+        If ``z_ratio`` is not of type `int` or `float`.
+
+    `ValueError`
+        If ``density`` is negative.
 
     Returns
     -------
-    ~astropy.units.Quantity
-        The mass density calculated from all the provided sources of information.
+    `~astropy.units.Quantity`
+        The mass density for the plasma species represented by ``particle``.
 
     Examples
     -------
-    >>> from astropy import units as u
-    >>> mass_density(1 * u.m ** -3,'p')
-    <Quantity 1.67353...e-27 kg / m3>
-    >>> mass_density(4 * u.m ** -3,'D+')
-    <Quantity 1.33779...e-26 kg / m3>
-
+    >>> import astropy.units as u
+    >>> mass_density(1 * u.m ** -3, 'p')
+    <Quantity 1.67262...e-27 kg / m3>
+    >>> mass_density(4 * u.m ** -3, 'D+')
+    <Quantity 1.33743...e-26 kg / m3>
+    >>> mass_density(2.e12 * u.cm ** -3, 'He')
+    <Quantity 1.32929...e-08 kg / m3>
+    >>> mass_density(2.e12 * u.cm ** -3, 'He', z_ratio=0.5)
+    <Quantity 6.64647...e-09 kg / m3>
+    >>> mass_density(1.0 * u.g * u.m ** -3, "")
+    <Quantity 0.001 kg / m3>
     """
-    # validate_quantities ensures we have units of u.kg/u.m**3 or 1/u.m**3
-    rho = density
-    if not rho.unit.is_equivalent(u.kg / u.m ** 3):
-        if particle:
-            m_i = particles.particle_mass(particle)
-            Z = _grab_charge(particle, z_mean)
-            rho = density * m_i + Z * density * m_e
-        else:
-            raise ValueError(
-                f"If passing a number density, you must pass a "
-                f"particle (not {particle}) to calculate the mass density!"
+    if density.unit.is_equivalent(u.kg / u.m ** 3):
+        return density
+
+    if not isinstance(particle, Particle):
+        try:
+            particle = Particle(particle)
+        except TypeError:
+            raise TypeError(
+                f"If passing a number density, you must pass a plasmapy Particle "
+                f"(not type {type(particle)}) to calculate the mass density!"
```

Comment:
We ca not do that at this point because the `density` annotation is a list.  That is what was causing the error that made me do this patch.  When I finally finish the update to `particle_input` that also adds `particle_like`, then we can replace this code block with `@particle_input`.

Reply:
Ah!  I hereby withdraw my suggestion.